rosys/vision/spatial_resection.py

```diff
@@ -0,0 +1,198 @@
+
+        if s0 is not None and isinstance(s0, list):
+            s0 = np.array(s0)
+
+        # Check sizes
+
+        assert len(world_lines) == len(image_line_points)
+        m = len(world_lines)
+
+        if world_points is None or image_points is None:
+            n = 0
+            world_points = self._DUMMY_ARRAY_3D
+        else:
+            assert len(world_points) == len(image_points)
+            n = len(world_points)
+
+        # Undistort observations
+
+        calibration = Calibration(intrinsics=intrinsics)
+        camera_matrix = calibration.get_undistorted_camera_matrix()
+
+        image_line_points = calibration.undistort_points(image_line_points)
+        if image_points is not None:
+            image_points = calibration.undistort_points(image_points)
+        else:
+            image_points = self._DUMMY_ARRAY_2D
+
+        all_image_points = np.concatenate([image_points, image_line_points], axis=0)
+
+        # Prepare state vector
+        # - 3D position of the camera
+        # - 4D quaternion of the camera rotation
+        # - m running variables for the lines
+        #  => [x, y, z, q_w, q_x, q_y, q_z, s_0, s_1, ..., s_m]
+
+        pos_0 = p0.array
+        q_0 = r0.quaternion
+        s_0 = s0 if s0 is not None else np.zeros(m)
+        x_0 = np.concatenate([pos_0.ravel(), q_0, s_0])
+
+        # Memory allocation (to avoid reallocations)
+        R = np.eye(3).astype(np.float64)
+        all_world_points = np.zeros(shape=(n + m, 3))
+        image_points_projected = np.zeros(shape=(n + m, 2))
+        residuals = np.zeros(shape=(n + m, 2))
+        world_line_points = np.zeros(shape=(m, 3))
+
+        def _quaternion_to_rotation_matrix(q: np.ndarray, R: np.ndarray) -> None:
+            """Inplace conversion of quaternion to rotation matrix.
+            q = [w, x, y, z]"""
+            R[0, 0] = 1 - 2*q[2]**2 - 2*q[3]**2
+            R[0, 1] = 2*q[1]*q[2] - 2*q[0]*q[3]
+            R[0, 2] = 2*q[1]*q[3] + 2*q[0]*q[2]
+            R[1, 0] = 2*q[1]*q[2] + 2*q[0]*q[3]
+            R[1, 1] = 1 - 2*q[1]**2 - 2*q[3]**2
+            R[1, 2] = 2*q[2]*q[3] - 2*q[0]*q[1]
+            R[2, 0] = 2*q[1]*q[3] - 2*q[0]*q[2]
+            R[2, 1] = 2*q[2]*q[3] + 2*q[0]*q[1]
+            R[2, 2] = 1 - 2*q[1]**2 - 2*q[2]**2
+
+        def f(x: np.ndarray) -> np.ndarray:
+            """Functional model using the collinearity equations."""
+
+            # Transform Object coordinates to Camera coordinates
+            C = x[:3]
+            q = x[3:7] / np.linalg.norm(x[3:7])
+            _quaternion_to_rotation_matrix(q, R)
+
+            np.multiply(world_lines[:, 3:], x[7:].reshape(-1, 1), out=world_line_points)
+            np.add(world_line_points, world_lines[:, :3], out=world_line_points)
+
+            all_world_points[:n] = world_points
+            all_world_points[n:] = world_line_points
+            np.subtract(all_world_points, C, out=all_world_points)
+
+            Xc = (R @ all_world_points.T).T
```

Comment:
should be the same as
```suggestion
            Xc = (all_world_points @ R.T)
```